File: xamarin/android/PropertyCross/Views/SearchResultsView.cs
using System;

using Android.App;
using Android.Content;
using Android.Content.PM;
using Android.Locations;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using Android.OS;
using PropertyFinder.Presenter;
using System.Collections.Generic;
using PropertyFinder.Model;
using Com.Actionbarsherlock.App;
using Com.Actionbarsherlock.View;

using IMenuItem = global::Com.Actionbarsherlock.View.IMenuItem;
using Android.Text;
using Android.Graphics;

namespace com.propertycross.xamarin.android.Views
{
	[Activity (ScreenOrientation = ScreenOrientation.Portrait)]	
	public class SearchResultsView : SherlockListActivity, SearchResultsPresenter.View
	{
		private View footer;
		private TextView resultDetails;
		private SearchResultsPresenter presenter;

		protected override void OnCreate(Bundle bundle)
		{
			base.OnCreate(bundle);

			SupportActionBar.SetDisplayHomeAsUpEnabled(true);

			LayoutInflater li = (LayoutInflater)GetSystemService(Context.LayoutInflaterService);
			footer = li.Inflate(Resource.Layout.load_more_footer, null);
			resultDetails = (TextView) footer.FindViewById(Resource.Id.result_details);
			footer.Click += OnFooterClicked;

			ListView.AddFooterView(footer);
			ListAdapter = new SearchResultsAdapter(this, new List<Property>() {});

			var app = PropertyFinderApplication.GetApplication(this);
			presenter = (SearchResultsPresenter) app.Presenter;
			presenter.SetView(this);
			app.CurrentActivity = this;
		}

		public override bool OnOptionsItemSelected(IMenuItem item)
		{
			if(item.ItemId == Android.Resource.Id.Home)
			{
				Finish();
				return true;
			}
			return base.OnOptionsItemSelected(item);
		}

		public void SetSearchResults(int totalResult, int pageNumber, int totalPages,
        	List<Property> properties, string searchLocation)
        {
			// Format the text:
			// Results for x, showing y of z properties.
			String text = Java.Lang.String.Format(Resources.GetString(Resource.String.result_details),
			                                      searchLocation, properties.Count, totalResult);
			resultDetails.TextFormatted = Html.FromHtml (text);

			((SearchResultsAdapter) ListAdapter).AddRange(properties);
			SupportActionBar.Title = Java.Lang.String.Format(Resources.GetString(Resource.String.results_shown),
			                                                 properties.Count,
			                                                 totalResult);
        }
        
        public void SetLoadMoreVisible(bool visible)
        {
			/* List View footers cannot be hidden easily because they're wrapped
			 * in an enclosing View. Instead, remove the footer and add it back
			 * if necessary.
			 */
			ListView.RemoveFooterView(footer);
			if (visible)
			{
				ListView.AddFooterView(footer);
			}
        }

		public bool IsLoading
		{
			set
			{
				footer.Enabled = !value;
				if(value)
				{
					resultDetails.Text = Resources.GetString(Resource.String.loading);
				}
			}
		}

      	public event EventHandler LoadMoreClicked;
      	public event EventHandler<PropertyEventArgs> PropertySelected;

		protected override void OnListItemClick(ListView l, View v, int position, long id)
		{
			var adapter = (SearchResultsAdapter) ListAdapter;
			Property item = adapter[position];
			PropertySelected(this, new PropertyEventArgs(item));
		}

		private void OnFooterClicked(object sender, EventArgs e)
		{
			LoadMoreClicked(this, EventArgs.Empty);
		}
	}
}	

File: native/android/PropertyCross/src/com/propertycross/android/views/SearchResultsView.java
package com.propertycross.android.views;

import java.util.ArrayList;
import java.util.List;

import android.content.Context;
import android.graphics.Typeface;
import android.os.Bundle;
import android.text.Html;
import android.text.Spannable;
import android.text.style.StyleSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ListView;
import android.widget.TextView;

import com.actionbarsherlock.app.SherlockListActivity;
import com.actionbarsherlock.view.MenuItem;
import com.propertycross.android.R;
import com.propertycross.android.events.Callback;
import com.propertycross.android.events.PropertyEventArgs;
import com.propertycross.android.events.PropertySelectedEvent;
import com.propertycross.android.events.UIEvent;
import com.propertycross.android.model.Property;
import com.propertycross.android.presenter.SearchResultsPresenter;

public class SearchResultsView 
	extends SherlockListActivity
	implements SearchResultsPresenter.View {

	private View footer;
	private TextView resultDetails;
	private SearchResultsPresenter presenter;
	private Callback<UIEvent> loadMoreCallback;
	private Callback<PropertySelectedEvent> propertySelectedCallback;
	
	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		
		getSupportActionBar().setDisplayHomeAsUpEnabled(true);
		
		LayoutInflater li = (LayoutInflater) this.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
		footer = li.inflate(R.layout.load_more_footer, null);
		footer.setOnClickListener(new View.OnClickListener(){

			@Override
			public void onClick(View v) {
				if (loadMoreCallback != null){
					loadMoreCallback.complete(new UIEvent(this));
				}
			}
			
		});
		resultDetails = (TextView) footer.findViewById(R.id.result_details);
		
		getListView().addFooterView(footer);
		setListAdapter(new SearchResultsAdapter(this, new ArrayList<Property>()));
		
		PropertyFinderApplication app = PropertyFinderApplication.getApplication(this);
		presenter = (SearchResultsPresenter) app.presenter;
		presenter.setView(this);
		app.currentActivity = this;
	}
	
	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		if (item.getItemId() == android.R.id.home) {
			finish();
			return true;
		}
		
		return super.onOptionsItemSelected(item);
	}
	
	@Override
	protected void onListItemClick(ListView l, View v, int position, long id) {
		Property item = ((SearchResultsAdapter) getListAdapter()).getItem(position);
		if (propertySelectedCallback != null) {
			propertySelectedCallback.complete(
					new PropertySelectedEvent(this, new PropertyEventArgs(item)));
		}
	}

	@Override
	public void setSearchResults(int totalResult, int pageNumber,
			int totalPages, List<Property> properties, String searchLocation) {
		
		// Format the text:
		// Results for x, showing y of z properties.
		String text = String.format(getResources().getString(R.string.result_details),
				searchLocation, properties.size(), totalResult);
		resultDetails.setText(Html.fromHtml(text));
		
		((SearchResultsAdapter) getListAdapter()).addRange(properties);
		getSupportActionBar().setTitle(String.format(
				getResources().getString(R.string.results_shown),
				properties.size(), totalResult));
	}

	@Override
	public void setLoadMoreVisible(boolean isVisible) {
		/*
		 * List View footers cannot be hidden easily because they're wrapped
		 * in an enclosing View. Instead, remove the footer and add it back
		 * if necessary.
		 */
		getListView().removeFooterView(footer);
		if (isVisible)
		{
			getListView().addFooterView(footer);
		} 
	}

	@Override
	public void setIsLoading(boolean isLoading) {
		footer.setEnabled(!isLoading);
		if (isLoading) {
			resultDetails.setText(R.string.loading);
		}
	}

	@Override
	public void loadMoreClicked(Callback<UIEvent> callback) {
		loadMoreCallback = callback;
	}

	@Override
	public void propertySelected(Callback<PropertySelectedEvent> callback) {
		propertySelectedCallback = callback;
	}	
}